tough/src/editor/targets.rs

```diff
@@ -0,0 +1,536 @@
+        let reader = Box::new(fetch_max_size(
+            transport,
+            role_url,
+            limits.max_targets_size,
+            "max targets limit",
+        )?);
+        // Load incoming role metadata as Signed<Targets>
+        let role: Signed<crate::schema::Targets> =
+            serde_json::from_reader(reader).context(error::ParseMetadata {
+                role: RoleType::Targets,
+            })?;
+
+        // Create `Signed<DelegatedTargets>` for the role
+        let delegated_targets = Signed {
+            signed: DelegatedTargets {
+                name: name.to_string(),
+                targets: role.signed.clone(),
+            },
+            signatures: role.signatures.clone(),
+        };
+        let (keyids, key_pairs) = if let Some(keys) = keys {
+            (keys.keys().cloned().collect(), keys)
+        } else {
+            let key_pairs = role.signed.delegations.context(error::NoDelegations)?.keys;
+            (key_pairs.keys().cloned().collect(), key_pairs)
+        };
+
+        self.delegate_role(delegated_targets, paths, key_pairs, keyids, threshold)?;
+
+        Ok(self)
+    }
+
+    /// Build the `Targets` struct
+    /// Adds in the new roles and new targets
+    pub fn build_targets(&self) -> Result<DelegatedTargets> {
+        let version = self.version.context(error::Missing {
+            field: "targets version",
+        })?;
+        let expires = self.expires.context(error::Missing {
+            field: "targets expiration",
+        })?;
+
+        // BEWARE!!! We are allowing targets to be empty! While this isn't
+        // the most common use case, it's possible this is what a user wants.
+        // If it's important to have a non-empty targets, the object can be
+        // inspected by the calling code.
+        let mut targets: HashMap<String, Target> = HashMap::new();
+        if let Some(ref existing_targets) = self.existing_targets {
+            targets.extend(existing_targets.clone());
+        }
+        if let Some(ref new_targets) = self.new_targets {
+            targets.extend(new_targets.clone());
+        }
+
+        let mut delegations = self.delegations.clone();
+        if let Some(delegations) = delegations.as_mut() {
+            if let Some(new_roles) = self.new_roles.as_ref() {
+                delegations.roles.extend(new_roles.clone());
+            }
+        }
+
+        let _extra = self._extra.clone().unwrap_or_else(HashMap::new);
+        Ok(DelegatedTargets {
+            name: self.name.clone(),
+            targets: Targets {
+                spec_version: SPEC_VERSION.to_string(),
+                version,
+                expires,
+                targets,
+                _extra,
+                delegations,
+            },
+        })
+    }
+
+    /// Creates a `Signed<DelegatedTargets>` for this role using the provided keys
+    pub fn create_signed(&self, keys: &[Box<dyn KeySource>]) -> Result<Signed<DelegatedTargets>> {
```

Comment:
Let's clarify the difference between this method and `sign()` below. When would I use them?

Reply:
When would I want to use `Signed<DelegatedTargets>` vs. `SignedDelegatedTargets`?